DSharpPlus.Core/Rest/Ratelimiting/PollyRatelimitingPolicy.cs

```diff
@@ -0,0 +1,172 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+using System;
+using System.Collections.Concurrent;
+using System.Net;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+
+using Microsoft.Extensions.Caching.Memory;
+
+using Polly;
+
+namespace DSharpPlus.Core.Rest.Ratelimiting
+{
+    // main ratelimit handler
+    internal class PollyRatelimitingPolicy : AsyncPolicy<HttpResponseMessage>
+    {
+        // keeps track of endpoint -> bucket hash mapping
+        private readonly ConcurrentDictionary<string, string> endpointBuckets;
+
+        // global ratelimit bucket of 50 requests per second
+        private readonly RatelimitBucket globalBucket;
+
+        // one second, we use this a lot so don't reallocate every time
+        private static readonly TimeSpan oneSecond = TimeSpan.FromSeconds(1);
+
+        public PollyRatelimitingPolicy()
+        {
+            globalBucket = new(50, 50, DateTimeOffset.UtcNow + oneSecond, "global");
+            endpointBuckets = new();
+        }
+
+        // the main policy.
+        protected override async Task<HttpResponseMessage> ImplementationAsync(
+            Func<Context, CancellationToken, Task<HttpResponseMessage>> action,
+            Context context,
+            CancellationToken cancellationToken,
+            bool continueOnCapturedContext)
+        {
+            // ensure a valid endpoint was passed
+            if (!context.TryGetValue("endpoint", out object endpointRaw) || endpointRaw is not string endpoint)
+            {
+                throw new InvalidOperationException("No endpoint passed.");
+            }
+
+            // ensure a valid cache was passed. the cache mostly serves as an alternative to v4's "BucketCleanerTask",
+            // which, in its static manner, would hardly work with the very DI-focused architecture of Polly and Polly policies.
+            if (!context.TryGetValue("cache", out object cacheRaw) || cacheRaw is not IMemoryCache cache)
```

Comment:
Will need to be changed if we do plan to support out-of-process caching (e.g. redis)

Reply:
storing ratelimit buckets remotely... i agree in theory, will change when we deal with out-of-process caching